### PR TITLE
Introduce the `PushBytes` trait to avoid using `Vec` in `Codec`

### PR DESCRIPTION
**Pull request by pvdrz**

### CODE REVIEW

File: rustls/src/client/tls12.rs [hunk 1]
```diff
@@ -388,7 +388,10 @@ impl State<ClientConnectionData> for ExpectServerKx {
 
         // Save the signature and signed parameters for later verification.
         let mut kx_params = Vec::new();
-        ecdhe.params.encode(&mut kx_params);
+        ecdhe
```
Comment (djc) [inline rustls/src/client/tls12.rs:391]: The proliferation of `unwrap()` calls seems like a substantial downside to this approach, I don't think that's the right way to do it.

File: rustls/src/client/tls12.rs [hunk 2]
```diff
@@ -432,7 +435,7 @@ fn emit_certificate(
 fn emit_clientkx(transcript: &mut HandshakeHash, common: &mut CommonState, pub_key: &[u8]) {
     let mut buf = Vec::new();
     let ecpoint = PayloadU8::new(Vec::from(pub_key));
-    ecpoint.encode(&mut buf);
+    ecpoint.encode(&mut buf).unwrap();
```
Comment (ctz) [inline rustls/src/client/tls12.rs:438]: Could use `get_encoding`?

File: rustls/src/common_state.rs [hunk 3]
```diff
@@ -392,7 +392,7 @@ impl CommonState {
                         "QUIC uses TLS for the cryptographic handshake only"
                     );
                     let mut bytes = Vec::new();
-                    m.payload.encode(&mut bytes);
```
Comment (ctz) [inline rustls/src/common_state.rs:395]: Ditto, `get_encoding` candidate.

File: rustls/src/msgs/message.rs [hunk 4]
```diff
@@ -235,7 +239,7 @@ impl From<Message> for PlainMessage {
             MessagePayload::ApplicationData(payload) => payload,
             _ => {
                 let mut buf = Vec::new();
-                msg.payload.encode(&mut buf);
+                msg.payload.encode(&mut buf).unwrap();
```
Comment (ctz) [inline rustls/src/msgs/message.rs:242]: Ditto `get_encoding`

File: rustls/src/tls12/mod.rs [hunk 5]
```diff
@@ -327,7 +329,9 @@ mod tests {
         let key = X25519.start().unwrap();
         let server_params = ServerECDHParams::new(&*key);
         let mut server_buf = Vec::new();
-        server_params.encode(&mut server_buf);
+        server_params
+            .encode(&mut server_buf)
+            .unwrap();
         server_buf.push(34);
```
Comment (ctz) [inline rustls/src/tls12/mod.rs:335]: Ditto `get_encoding`

File: rustls/src/common_state.rs (hunk 3 above)
Comment (pvdrz) [reply rustls/src/common_state.rs]: can't do. `MessagePayload` has a custom `encode` method instead of implementing `Codec`.

File: rustls/src/client/tls12.rs (hunk 1 above)
Comment (pvdrz) [reply rustls/src/client/tls12.rs]: I pushed some changes to avoid adding those extra `unwrap()`s

File: rustls/src/client/tls12.rs [hunk 6]
```diff
@@ -430,10 +428,7 @@ fn emit_certificate(
 }
 
 fn emit_clientkx(transcript: &mut HandshakeHash, common: &mut CommonState, pub_key: &[u8]) {
-    let mut buf = Vec::new();
-    let ecpoint = PayloadU8::new(Vec::from(pub_key));
-    ecpoint.encode(&mut buf);
-    let pubkey = Payload::new(buf);
+    let pubkey = Payload::new(PayloadU8::new(Vec::from(pub_key)).get_encoding());
```
Comment (djc) [inline rustls/src/client/tls12.rs:431]: Please avoid making unrelated changes in the same PR or commit.

File: rustls/src/msgs/alert.rs [hunk 7]
```diff
@@ -12,9 +12,9 @@ pub struct AlertMessagePayload {
 }
 
 impl Codec for AlertMessagePayload {
-    fn encode(&self, bytes: &mut Vec<u8>) {
-        self.level.encode(bytes);
-        self.description.encode(bytes);
+    fn try_encode<B: TryPushBytes>(&self, bytes: &mut B) -> Result<(), B::Error> {
```
Comment (djc) [inline rustls/src/msgs/alert.rs:15]: I think we should avoid renaming the method in the same commit (or, in the current state, PR) as change the type, since this creates a bunch of otherwise unrelated churn.

File: rustls/src/client/tls12.rs [hunk 8]
```diff
@@ -387,9 +387,7 @@ impl State<ClientConnectionData> for ExpectServerKx {
             })?;
 
         // Save the signature and signed parameters for later verification.
-        let mut kx_params = Vec::new();
-        ecdhe.params.encode(&mut kx_params);
-        let server_kx = ServerKxDetails::new(kx_params, ecdhe.dss);
+        let server_kx = ServerKxDetails::new(ecdhe.params.get_encoding(), ecdhe.dss);
```
Comment (djc) [inline rustls/src/client/tls12.rs:390]: Would be nice to land these changes separately (either in a separate commit if you're going to squash and clean up commits here or in a separate PR).

File: rustls/src/client/tls12.rs (hunk 6 above)
Comment (pvdrz) [reply rustls/src/client/tls12.rs:431]: Oh I wasn't to but @ctz suggested it on their review. I'll revert it then.

File: rustls/src/msgs/alert.rs (hunk 7 above)
Comment (pvdrz) [reply rustls/src/msgs/alert.rs:15]: I think I don't follow. `Codec::encode` is still used as it was before, it's signature changed but it is backwards compatible with the old one.

File: rustls/src/client/tls12.rs (hunk 8 above)
Comment (pvdrz) [reply rustls/src/client/tls12.rs:390]: Same answer as https://github.com/rustls/rustls/pull/1533#discussion_r1357165250

File: rustls/src/common_state.rs (hunk 3 above)
Comment (ctz) [reply rustls/src/common_state.rs]: Fair enough!

File: rustls/src/msgs/alert.rs (hunk 7 above)
Comment (djc) [reply rustls/src/msgs/alert.rs:15]: Ahh, sorry, yes, I see what you did now.

**Comment by djc** [comment]
I think this is essentially the right direction but I also think it should get a bunch of polish.

* Naming: the trait and type names you're introducing have very low-level, mechanic names. I think it would be good to use names that evoke more concept-level meaning. For example, `ByteBuffer` is an extremely generic name, whereas the specific functionality that it is trying to facilitate is fairly specific (to allow single-pass encoding of length-prefixed data). Similarly, `PushBytes` could be more like `ResizableBuffer`. It would also be good to add comments to each of the traits and types introduced that explain what they're trying to achieve (for example, I'm still unclear on what `SliceBuffer` is for).
* Ordering: in the rustls project we try to maintain a [top-down ordering](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md#top-down-ordering-within-modules), so dependent types/traits should come before lower-level types/traits in the source. This is not yet followed everywhere but would be good to apply for the new code here, especially since it's a pretty complicated hierarchy of stuff.
* Commit history: currently the commit history is pretty much a mess. It would help a lot if there are ways to split these changes into smaller chunks. For example, I think the `ByteBuffer` trait could be introduced before the other stuff.

**Comment by cpu** [comment]
@pvdrz Was this closed because it's no longer necessary, or for some other reason?

**Comment by pvdrz** [comment]
Yeah this was superseded by #1595 basically
